src/repr/src/adt/date.rs

```diff
@@ -0,0 +1,178 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A time Date abstract data type.
+
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Sub;
+
+use anyhow::anyhow;
+use chrono::NaiveDate;
+use mz_proto::{RustType, TryFromProtoError};
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+include!(concat!(env!("OUT_DIR"), "/mz_repr.adt.date.rs"));
+
+#[derive(Debug, Error)]
+pub enum DateError {
+    #[error("data out of range")]
+    OutOfRange,
+}
+
+/// A Postgres-compatible Date. Additionally clamp valid dates for the range
+/// that chrono supports to allow for safe string operations. Infinite dates are
+/// not yet supported.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Hash, Deserialize)]
+pub struct Date {
+    /// Number of days from the postgres epoch (2000-01-01).
+    days: i32,
+}
+
+impl RustType<ProtoDate> for Date {
+    fn into_proto(&self) -> ProtoDate {
+        ProtoDate { days: self.days }
+    }
+
+    fn from_proto(proto: ProtoDate) -> Result<Self, TryFromProtoError> {
+        Ok(Date { days: proto.days })
+    }
+}
+
+impl std::str::FromStr for Date {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        crate::strconv::parse_date(s).map_err(|e| anyhow!(e))
+    }
+}
+
+static PG_EPOCH: Lazy<NaiveDate> = Lazy::new(|| NaiveDate::from_ymd(2000, 1, 1));
+
+impl Date {
+    pub const UNIX_EPOCH_TO_PG_EPOCH: i32 = 10957; // Number of days from 1970-01-01 to 2000-01-01.
+    const CE_EPOCH_TO_PG_EPOCH: i32 = 730120; // Number of days since 0001-01-01 to 2000-01-01.
+    pub const LOW_DAYS: i32 = -2451545; // 4714-11-24 BC
+
+    /// Largest date support by Materialize. Although Postgres can go up to
+    /// 5874897-12-31, chrono is limited to December 31, 262143, which we mirror
+    /// here so we can use chrono's formatting methods and have guaranteed safe
+    /// conversions.
+    pub const HIGH_DAYS: i32 = 95_015_644;
+
+    /// Constructs a new `Date` as the days since the postgres epoch
+    /// (2000-01-01).
+    pub fn from_pg_epoch(days: i32) -> Result<Date, DateError> {
+        if days < Self::LOW_DAYS || days > Self::HIGH_DAYS {
+            Err(DateError::OutOfRange)
+        } else {
+            Ok(Date { days })
+        }
+    }
+
+    /// Constructs a new `Date` as the days since the Unix epoch.
+    pub fn from_unix_epoch(unix_days: i32) -> Result<Date, DateError> {
+        let pg_days = unix_days.saturating_sub(Self::UNIX_EPOCH_TO_PG_EPOCH);
+        if pg_days == i32::MIN {
+            return Err(DateError::OutOfRange);
+        }
+        Self::from_pg_epoch(pg_days)
+    }
+
+    /// Returns the number of days since the postgres epoch.
+    pub fn pg_epoch_days(&self) -> i32 {
+        self.days
+    }
+
+    pub fn is_finite(&self) -> bool {
```

Comment:
Blergh I swapped the logic here. MAX and MIN are the infinite dates (see https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-INPUT section 8.5.1.4). We don't support them yet, but I wanted to add in some protection if we ever do because some stuff would treat those sentinel values incorrectly.

Reply:
Fixed, and added a comment.